### ISSUE TITLE
Consider using different Gradle attributes for `dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable`

### ISSUE THREAD

**Issue by martinbonnin** [adamko-dev/dokkatoo#165]
After upgrating to `2.1.0`, we started having KSP errors because the KSP classpath would select the `dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable` variant of some projects instead of the `jvmRuntimeElements` ones. They are indeed quite similar:

```
--------------------------------------------------
Variant dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable
--------------------------------------------------
Shared Dokka Plugins for consumers that will assemble a html Publication using the Dokka Module that this project produces

Capabilities
    - com.apollographql.apollo3:apollo-ast:4.0.0-beta.5-SNAPSHOT (default capability)
Attributes
    - dev.adamko.dokkatoo.classpath  = dokka-publication-plugins
    - dev.adamko.dokkatoo.format     = html
    - org.gradle.category            = library
    - org.gradle.dependency.bundling = external
    - org.gradle.jvm.environment     = standard-jvm
    - org.gradle.libraryelements     = jar
    - org.gradle.usage               = java-runtime
```

vs

```
--------------------------------------------------
Variant jvmRuntimeElements
--------------------------------------------------
Elements of runtime for main.

Capabilities
    - com.apollographql.apollo3:apollo-ast:4.0.0-beta.5-SNAPSHOT (default capability)
Attributes
    - org.gradle.category                = library
    - org.gradle.jvm.environment         = standard-jvm
    - org.gradle.libraryelements         = jar
    - org.gradle.usage                   = java-runtime
    - org.jetbrains.kotlin.platform.type = jvm
Artifacts
    - build/libs/apollo-ast-jvm-4.0.0-beta.5-SNAPSHOT.jar (artifactType = jar)
```

I was able to workaround by forcing "unselection" of the dokkatoo variant:

```kotlin
  configurations.configureEach {
    if (name.lowercase().contains("dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable".lowercase())) {
      attributes {
        attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage::class.java, "poison"))
      }
    }
  }
```

TBH I'm not 100% what the root cause is but would it be possible to use more specific attributes for `dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable`? Maybe `org.gradle.usage = dokkatoo-plugin`?

**Comment by aSemy** [comment]
Thanks for the report!

The configurations have different attributes so it's a mystery why Gradle is confusing them. 

I've tried various combinations of setting `org.gradle.usage` but nothing seems to work. It needs to be set to 'java-runtime', otherwise Gradle won't fetch the dependency from Maven Central.

It looks like this is _another_ bug/flaw/undocumented feature in Gradle so I've raised an issue https://github.com/gradle/gradle/issues/28053. I'm not sure how best to workaround it.

**Comment by martinbonnin** [comment]
Smaller reproducer with included builds instead of KSP here: https://github.com/martinbonnin/reproducer-attributes

tldr; not really sure what's happening down there but given all the existing ecosystem around this, I'd feel more comfortable if the dokkattoo artifacts were using a dedicated `org.gradle.usage` attribute. Maybe there need to be an "intermediate" configuration? One that resolves from Maven Central using `java-runtime` but then exposes something else in outgoingVariants?

**Comment by aSemy** [comment]
This issue seems relevant https://github.com/gradle/gradle/issues/26463

**Comment by aSemy** [comment]
#173 seems to have helped in some cases, but not all

https://github.com/apollographql/apollo-kotlin/commit/a58454c7d3f3bfcd7ea2fd64938b10f69571d402

**Comment by martinbonnin** [comment]
Right. It's working when everything is in the same build but failing to compile build scripts in composite builds. From that commit `./gradlew -p tests build` fails IIRC

**Comment by aSemy** [comment]
hey @martinbonnin, when you try K2 could you also try and remove the 'poison' workaround? I'm curious whether it would help. See https://github.com/adamko-dev/dokkatoo/issues/170#issuecomment-2094210736

**Comment by martinbonnin** [comment]
Not working for us sadly ([commit](https://github.com/apollographql/apollo-kotlin/commit/a58454c7d3f3bfcd7ea2fd64938b10f69571d402)):

```
> Configure project :compiler-plugins-app
w: file:///Users/mbonnin/git/apollo-kotlin/tests/compiler-plugins/app/build.gradle.kts:1:1: @OptIn without any arguments has no effect
e: file:///Users/mbonnin/git/apollo-kotlin/tests/compiler-plugins/app/build.gradle.kts:1:13: Unresolved reference: ApolloExperimental
e: file:///Users/mbonnin/git/apollo-kotlin/tests/compiler-plugins/app/build.gradle.kts:1:13: An annotation argument must be a compile-time constant
e: file:///Users/mbonnin/git/apollo-kotlin/tests/compiler-plugins/app/build.gradle.kts:3:34: Unresolved reference: annotations
```

This somewhat fails when resolving included build. Not that this commit is using KGP RC2 but languageVersion 1.9 because we're stuck on other issues to enable languageVersion 2.0 (but I'd say the important thing is the KGP version here?)

**Comment by aSemy** [comment]
Thanks for trying! Did you also check with the latest Dokkatoo version? The [commit you linked](https://github.com/apollographql/apollo-kotlin/compare/update-dokkatoo#diff-35f86fd5014744e66d7d959455aa7285ccc0d92e1c8677d670ea6308f5afcb4bR97) shows v2.2.0, but the latest is v2.3.1. This contains a fix relevant for composite builds - https://github.com/gradle/gradle/issues/28695

**Comment by martinbonnin** [comment]
Wooops, apologies I didn't realize there was an old branch named `update-dokkatoo` and my push didn't go through. 

[This is the good commit](https://github.com/apollographql/apollo-kotlin/compare/main...update-dokkatoo) I tried yesterday. It still fails with `2.3.1` sadly.

**Comment by Vampire** [comment]
I didn't look at the commits posted here in the last couple of comments, but had a look at the original reproducer from https://github.com/adamko-dev/dokkatoo/issues/165#issuecomment-1942382698 as is pushed there right now.

The problem is, that `jvmRuntimeElements` is missing the `Bundling` attribute.
This is requested:
```plain
    | org.gradle.category                |  library                   |
    | org.gradle.dependency.bundling     |  external                  |
    | org.gradle.libraryelements         |  jar                       |
    | org.gradle.usage                   |  java-runtime              |
```
`javaRuntimeElements` provides
```plain
    | org.gradle.category                | library                   |
    | org.gradle.dependency.bundling     |                           |
    | org.gradle.libraryelements         | jar                       |
    | org.gradle.usage                   | java-runtime              |

    | org.gradle.jvm.environment         | standard-jvm              |
    | org.jetbrains.kotlin.platform.type | jvm                       |
```
`dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable` provides
```plain
    | org.gradle.category                | library                   |
    | org.gradle.dependency.bundling     | external                  |
    | org.gradle.libraryelements         | jar                       |
    | org.gradle.usage                   | java-runtime              |

    | org.gradle.jvm.environment         | standard-jvm              |
    | dev.adamko.dokkatoo.classpath      | dokka-publication-plugins |
    | dev.adamko.dokkatoo.format         | html                      |
```

So `dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable` simply matches better than `jvmRuntimeElements` as it matches 4 of the 3 requested attributes while the latter only matches 3.

So the first bug here is, that the `jvmRuntimeElements` configuration does not provide the JVM ecosystem `Bundling` attribute.

If now that variant would also provide the `Bundling` attribute,
both variants would provide the same matching attributes but have different extra attributes.
You can emulate this by adding to the `annotations` build script
```kotlin
configurations.matching { it.name == "jvmRuntimeElements" }.configureEach {
    attributes {
        attribute(Bundling.BUNDLING_ATTRIBUTE, objects.named(Bundling.EXTERNAL))
    }
}
```
With that it will then fail to resolve the dependency as it cannot decide between those two configurations,
becuase both have the extra attribute `org.gradle.jvm.environment` with value `standard-jvm`,
but one has the additional extra attribute `org.jetbrains.kotlin.platform.type`
while the other has the extra attributes `dev.adamko.dokkatoo.classpath` and `dev.adamko.dokkatoo.format`
and Gradle does not know which would be better.

I'd say the next bug is, that actually the `dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable` does not have `org.jetbrains.kotlin.platform.type` with value `jvm`,
so let's also add that by also in the `annotations` project doing
```kotlin
configurations.matching { it.name == "dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable" }.configureEach {
    attributes {
        attribute(KotlinPlatformType.attribute, KotlinPlatformType.jvm)
    }
}
```

Now both configurations have the same matching attributes.
Both have the same 2 extra attributes with the same value.
One has no more extra attributes.
The other has two more extra attributes unique to them.

Now the attribute resolution algorithm described at https://docs.gradle.org/current/userguide/variant_aware_resolution.html#sec:abm-algorithm 
can properly resolve the intended configuration and the annotation is found as expected.

So I'd say the bugs to fix are, that the `jvmRuntimeElements` configuration is missing the `Bundling` attribute,
and that the `dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable` configuration is missing the `KotlinPlatformType.attribute` attribute.

Once these two are fixed, there is no hacky work-around with Dokka-own attributes and compatibility rules, or poison values necessary,
as then the resolution just works as intended.

**Comment by martinbonnin** [comment]
Thanks for diving in. I agree 100% with your analysis and the fact we can use the `Candidates that lack extra attributes are preferred.` part of the algorithm to solve this.

That being said, the fact that `jvmRuntimeElements` is missing the `Bundling` attribute is probably on the KMP plugin. Asking every build to add the attribute is not scalable.

We could probably fix it in the KMP plugin itself but nothing prevents someone else to redo the same mistake. Something like [shadow](https://github.com/GradleUp/shadow/issues/1540) for an example exposes `java-runtime` outgoing variants and I wouldn't be surprised if some of the attributes would be missing.

Another option is to have dokka add the bundling attribute automatically but that feels dangerous because dokka has no knowledge about the bundling of variants provided by other plugins.

All in all, I stand by my [proposed solution](https://github.com/Kotlin/dokka/pull/4286) that adds an extra `org.jetbrains.dokka.classpath: none` attribute the `jvmRuntimeElements`. It's more explicit that the attribute was added by dokka and easier to debug if (when) things go wrong.

**Comment by Vampire** [comment]
> That being said, the fact that jvmRuntimeElements is missing the Bundling attribute is probably on the KMP plugin. Asking every build to add the attribute is not scalable.

Of course, definitely. 
Did for sure not want to imply that builds should add that manually, that was just to ad-hoc test and show the outcome. 

Actually, the plugin adding `dokkatooHtmlPublicationPluginClasspathApiOnlyConsumable` could also depending on KMP version add the missing `Bundling` attribute to `jvmRuntimeElements`, or could maybe copy the `jvmRuntimeElements` attributes, adding the two additional ones, but no idea whether that would be a good idea.

> We could probably fix it in the KMP plugin itself

Yes, that was what I meant. :-)

> but nothing prevents someone else to redo the same mistake.

Of course, it can always someone somewhere introduce a bug. And then people stumble over them sooner or later and then they can be fixed. :-)

Especially, setting attributes, or even more defining new attributes, is not a light topic. Defining and maintaining attributes and their meaning and compatibility and disambiguation is actually one of the hardest topics, especially as the attributes make it into the published metadata.

> Something like shadow for an example exposes java-runtime outgoing variants and I wouldn't be surprised if some of the attributes would be missing.

Why? 
Do you think the initial developer or the maintainers don't know what they do? :-)
I personally generally don't like the concept of `shadow`, and no software is bug-free, but that does not mean that the bugs cannot get fixed or that you should always expect others to make wrong what could be done wrong. :-)
https://github.com/GradleUp/shadow/blob/4ff97b3e74a35d19f9e50ecb2088e2d1be8e20c0/src/main/kotlin/com/github/jengelman/gradle/plugins/shadow/ShadowJavaPlugin.kt#L57 looks promising from a quick peek.

> All in all, I stand by my proposed solution that adds an extra org.jetbrains.dokka.classpath: none attribute the jvmRuntimeElements. It's more explicit that the attribute was added by dokka and easier to debug if (when) things go wrong.

Well, but it is also questionable whether Dokka should add a "random" attribute to a configuration it does not "own". That could in other situations then cause problems, for example if before the "one has extra attributes" rule worked, then you apply Dokka plugin and suddenly both configurations have extra attributes each and resolution falls like when just applying the first snippet of my last comment. 

My gut feeling still indicates the two configurations should just be fixed by adding the missing ecosystem attributes. Everything else seems more a hack and work-around that should be done in affected builds, not a central plugin. 🤷‍♂️
